# IPython: Beyond Normal Python

    Help and Documentation in IPython
    Keyboard Shortcuts in the IPython Shell
    IPython Magic Commands
    Input and Output History
    IPython and Shell Commands
    Errors and Debugging
    Profiling and Timing Code

## Help and Documentation in IPython

One of the most useful functions of IPython/Jupyter is to shorten the gap between the user and the type of documentation and search that will help them do their work effectively. While web searches still play a role in answering complicated questions, an amazing amount of information can be found through IPython alone. Some examples of the questions IPython can help answer in a few keystrokes:

>How do I call this function? What arguments and options does it have?

>What does the source code of this Python object look like?

>What is in this package I imported? What attributes or methods does this object have?

Here we'll discuss IPython's tools to quickly access this information, namely the ? character to explore documentation, the ?? characters to explore source code, and the Tab key for auto-completion.

Python has a built-in **help()** function that can access this information and prints the results. For example, to see the documentation of the built-in len function, you can do the following:

In [1]:
help(len)

Help on built-in function len in module builtins:

len(obj, /)
    Return the number of items in a container.



Because finding help on an object is so common and useful, IPython introduces the **?** character as a shorthand for accessing this documentation and other relevant information:

In [3]:
len?

**Signature:** len(obj, /)

**Docstring:** Return the number of items in a container.

**Type:** builtin_function_or_method

In [4]:
L = [1,2,3]
L.insert?

**Docstring:** L.insert(index, object) -- insert object before index

**Type:** builtin_function_or_method

In [5]:
L?

**Type:**        list

**String form:** [1, 2, 3]

**Length:**      3

**Docstring:** 
list() -> new empty list
list(iterable) -> new list initialized from iterable's items

Importantly, this will even work for functions or other objects you create yourself! Here we'll define a small function with a docstring:

In [6]:
def square(a):
    """Return the square of a."""
    return a**2

In [7]:
square?

**Signature:** square(a)
    
**Docstring:** Return the square of a.
    
**File:**      ~/library/<ipython-input-6-b9a5499fe5b8>

**Type:**      function

This quick access to documentation via docstrings is one reason you should get in the habit of always adding such inline documentation to the code you write!

Because the Python language is so easily readable, another level of insight can usually be gained by reading the source code of the object you're curious about. IPython provides a shortcut to the source code with the double question mark (**??**):

In [8]:
square??

**Signature:** square(a)

**Source:   **
def square(a):

    """Return the square of a."""
    
    return a\*\*2
    
    
**File:**      ~/library/<ipython-input-6-b9a5499fe5b8>

**Type:**      function

If you play with this much, you'll notice that sometimes the **??** suffix doesn't display any source code: this is generally because the object in question is not implemented in Python, but in C or some other compiled extension language. If this is the case, the **??** suffix gives the same output as the **?** suffix.

In [9]:
len??

**Signature:** len(obj, /)

**Docstring:** Return the number of items in a container.

**Type:**      builtin_function_or_method

Every Python object has various attributes and methods associated with it. Like with the help function discussed before, Python has a built-in **dir** function that returns a list of these, but the tab-completion interface is much easier to use in practice. 

To see a list of all available attributes of an object, you can type the name of the object followed by a period (".") character and the Tab key:

In [10]:
L.pop #neat

<function list.pop>

To narrow-down the list, you can type the first character or several characters of the name, and the Tab key will find the matching attributes and methods.

If there is only a single option, pressing the Tab key will complete the line for you. 

For clarity, private methods and special methods are omitted from the list by default, but it's possible to list them by explicitly typing the underscore.

In [11]:
L.__add__

<method-wrapper '__add__' of list object at 0x7f8894500748>

Most of these are Python's special double-underscore methods (often nicknamed **"dunder"** methods).

Tab completion is also useful when importing objects from packages. We can use it to find all possible imports in the itertools package that start with co.

Similarly, you can use tab-completion to see which imports are available on your system.

Tab completion is useful if you know the first few characters of the object or attribute you're looking for, but is little help if you'd like to match characters at the middle or end of the word. For this use-case, IPython provides a means of wildcard matching for names using the __*__ character.

For example, we can use this to list every object in the namespace that ends with Warning:

In [12]:
*Warning?

BytesWarning

DeprecationWarning

FutureWarning

ImportWarning

PendingDeprecationWarning

ResourceWarning

RuntimeWarning

SyntaxWarning

UnicodeWarning

UserWarning

Warning

Notice that the * character matches any string, including the empty string.

Similarly, suppose we are looking for a string method that contains the word find somewhere in its name. We can search for it this way:

In [13]:
str.*find*?

str.find

str.rfind

## IPython Magic Commands

**Magic commands** are designed to succinctly solve various common problems in standard data analysis. Magic commands come in two flavors: line magics, which are denoted by a single **%** prefix and operate on a single line of input, and cell magics, which are denoted by a double **%%** prefix and operate on multiple lines of input. 

When working in the IPython interpreter, one common gotcha is that pasting multi-line code blocks can lead to unexpected errors, especially when indentation and interpreter markers are involved.

The code is formatted as it would appear in the Python interpreter, and if you copy and paste this directly into IPython you get an error:

In [16]:
>>> def donothing(x):
...     return x

In the direct paste, the interpreter is confused by the additional prompt characters. But never fear–IPython's **%paste** magic function is designed to handle this exact type of multi-line, marked-up input:

In [18]:
donothing(5) #this seems to work without the magic function

5

As you begin developing more extensive code, you will likely find yourself working in both IPython for interactive exploration, as well as a text editor to store code that you want to reuse. Rather than running this code in a new window, it can be convenient to run it within your IPython session. This can be done with the **%run** magic.

In [23]:
%run myscript.py

1 squared is 1
2 squared is 4
3 squared is 9


Note also that after you've run this script, any functions defined within it are available for use in your IPython session:

In [25]:
square(5)

25

There are several options to fine-tune how your code is run; you can see the documentation in the normal way, by typing %run? in the IPython interpreter.

In [6]:
%run?

Another example of a useful magic function is **%timeit**, which will automatically determine the execution time of the single-line Python statement that follows it. For example, we may want to check the performance of a list comprehension:

In [27]:
%timeit L = [n ** 2 for n in range(1000)]

334 µs ± 37.6 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


The benefit of **%timeit** is that for short commands it will automatically perform multiple runs in order to attain more robust results. For multi line statements, adding a second % sign will turn this into a cell magic that can handle multiple lines of input. For example, here's the equivalent construction with a for-loop:

In [28]:
%%timeit

L = []

for n in range(1000):
    L.append(n ** 2)

389 µs ± 17 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


To access a general description of available magic functions, including some examples, you can type this:

In [29]:
%magic

For a quick and simple list of all available magic functions, type this:

In [30]:
%lsmagic

Available line magics:
%alias  %alias_magic  %autocall  %automagic  %autosave  %bookmark  %cat  %cd  %clear  %colors  %config  %connect_info  %cp  %debug  %dhist  %dirs  %doctest_mode  %ed  %edit  %env  %gui  %hist  %history  %killbgscripts  %ldir  %less  %lf  %lk  %ll  %load  %load_ext  %loadpy  %logoff  %logon  %logstart  %logstate  %logstop  %ls  %lsmagic  %lx  %macro  %magic  %man  %matplotlib  %mkdir  %more  %mv  %notebook  %page  %pastebin  %pdb  %pdef  %pdoc  %pfile  %pinfo  %pinfo2  %popd  %pprint  %precision  %profile  %prun  %psearch  %psource  %pushd  %pwd  %pycat  %pylab  %qtconsole  %quickref  %recall  %rehashx  %reload_ext  %rep  %rerun  %reset  %reset_selective  %rm  %rmdir  %run  %save  %sc  %set_env  %store  %sx  %system  %tb  %time  %timeit  %unalias  %unload_ext  %who  %who_ls  %whos  %xdel  %xmode

Available cell magics:
%%!  %%HTML  %%SVG  %%bash  %%capture  %%debug  %%file  %%html  %%javascript  %%js  %%latex  %%markdown  %%perl  %%prun  %%pypy  %%python  %%python

Finally, I'll mention that it is quite straightforward to define your own magic functions if you wish.

## Input and Output History

The **In** object is a list, which keeps track of the commands in order (the first item in the list is a place-holder so that In[1] can refer to the first command):

In [31]:
In[1]

'help(len)'

The **Out** object is not a list but a dictionary mapping input numbers to their outputs (if any):

In [33]:
print(Out[10])

<built-in method pop of list object at 0x7f8894500748>


Note that not all operations have outputs: for example, import statements and print statements don't affect the output. 

The standard Python shell contains just one simple shortcut for accessing previous output; the variable _ (i.e., a single underscore) is kept updated with the previous output.

In [34]:
print(_)

help(len)


But IPython takes this a bit further—you can use a double underscore to access the second-to-last output, and a triple underscore to access the third-to-last output (skipping any commands with no output):

In [35]:
print(__)
print(___)

Available line magics:
%alias  %alias_magic  %autocall  %automagic  %autosave  %bookmark  %cat  %cd  %clear  %colors  %config  %connect_info  %cp  %debug  %dhist  %dirs  %doctest_mode  %ed  %edit  %env  %gui  %hist  %history  %killbgscripts  %ldir  %less  %lf  %lk  %ll  %load  %load_ext  %loadpy  %logoff  %logon  %logstart  %logstate  %logstop  %ls  %lsmagic  %lx  %macro  %magic  %man  %matplotlib  %mkdir  %more  %mv  %notebook  %page  %pastebin  %pdb  %pdef  %pdoc  %pfile  %pinfo  %pinfo2  %popd  %pprint  %precision  %profile  %prun  %psearch  %psource  %pushd  %pwd  %pycat  %pylab  %qtconsole  %quickref  %recall  %rehashx  %reload_ext  %rep  %rerun  %reset  %reset_selective  %rm  %rmdir  %run  %save  %sc  %set_env  %store  %sx  %system  %tb  %time  %timeit  %unalias  %unload_ext  %who  %who_ls  %whos  %xdel  %xmode

Available cell magics:
%%!  %%HTML  %%SVG  %%bash  %%capture  %%debug  %%file  %%html  %%javascript  %%js  %%latex  %%markdown  %%perl  %%prun  %%pypy  %%python  %%python

Python stops there: more than three underscores starts to get a bit hard to count, and at that point it's easier to refer to the output by line number.

There is one more shortcut we should mention, however–a shorthand for **Out[X]** is **\_X** (i.e., a single underscore followed by the line number):

In [36]:
_10

<function list.pop>

Sometimes you might wish to suppress the output of a statement.

r maybe the command you're executing produces a result that you'd prefer not like to store in your output history, perhaps so that it can be deallocated when other references are removed. The easiest way to suppress the output of a command is to add a semicolon to the end of the line:

In [37]:
import math

math.sin(2) + math.cos(2);

For accessing a batch of previous inputs at once, the %history magic command is very helpful. Here is how you can print the first four inputs:

In [38]:
%history -n 1-4

   1: help(len)
   2: len?
   3: len?
   4:
L = [1,2,3]
L.insert?


When working interactively with the standard Python interpreter, one of the frustrations is the need to switch between multiple windows to access Python tools and system command-line tools. IPython bridges this gap, and gives you a syntax for executing shell commands directly from within the IPython terminal. The magic happens with the exclamation point: anything appearing after **!** on a line will be executed not by the Python kernel, but by the system command-line.

Any command that works at the command-line can be used in IPython by prefixing it with the ! character. For example, the ls, pwd, and echo commands can be run as follows:

In [39]:
!ls
!pwd
!echo 'hey'

README.md  beyond_normal_python.ipynb  myscript.py
/home/nbuser/library
hey


Shell commands can not only be called from IPython, but can also be made to interact with the IPython namespace. For example, you can save the output of any shell command to a Python list using the assignment operator:

In [40]:
contents = !ls
directory = !pwd

In [41]:
contents

['README.md', 'beyond_normal_python.ipynb', 'myscript.py']

In [42]:
directory

['/home/nbuser/library']

This looks and acts a lot like a Python list, but has additional functionality, such as the grep and fields methods and the s, n, and p properties that allow you to search, filter, and display the results in convenient ways.

Communication in the other direction–passing Python variables into the shell–is possible using the {varname} syntax:

In [43]:
message = "hello from Python"

In [44]:
!echo {message}

hello from Python


f you play with IPython's shell commands for a while, you might notice that you cannot use !cd to navigate the filesystem:

The reason is that shell commands in the notebook are executed in a temporary subshell. If you'd like to change the working directory in a more enduring way, you can use the **%cd** magic command:.

Besides %cd, other available shell-like magic functions are **%cat, %cp, %env, %ls, %man, %mkdir, %more, %mv, %pwd, %rm, and %rmdir**, any of which can be used without the % sign if automagic is on. 

In [45]:
mkdir tmp

In [46]:
ls

README.md  beyond_normal_python.ipynb  myscript.py  tmp/


In [47]:
cp myscript.py tmp/

In [48]:
ls tmp

myscript.py


In [49]:
rm -r tmp

In [50]:
ls

README.md  beyond_normal_python.ipynb  myscript.py


## Errors and Debugging

Most of the time when a Python script fails, it will raise an Exception. When the interpreter hits one of these exceptions, information about the cause of the error can be found in the traceback, which can be accessed from within Python. 

With the **%xmode** magic function, IPython allows you to control the amount of information printed when the exception is raised.

In [64]:
%xmode Context

def func1(a,b):
    return a/b

def func2(x):
    a = x
    b = x - 1
    return func1(a, b)

Exception reporting mode: Context


In [65]:
func2(1)

ZeroDivisionError: division by zero

%xmode takes a single argument, the mode, and there are three possibilities: **Plain**, **Context**, and **Verbose**. The default is Context, and gives the output above. 

Plain is more compact and gives less information:

In [66]:
%xmode Plain

Exception reporting mode: Plain


In [67]:
func2(1)

ZeroDivisionError: division by zero

The Verbose mode adds some extra information, including the arguments to any functions that are called:

In [68]:
%xmode Verbose

Exception reporting mode: Verbose


In [70]:
func2(1)

ZeroDivisionError: division by zero

The standard Python tool for interactive debugging is **pdb**, the Python debugger. This debugger lets the user step through the code line by line in order to see what might be causing a more difficult error. The IPython-enhanced version of this is **ipdb**, the IPython debugger.

In IPython, perhaps the most convenient interface to debugging is the **%debug** magic command. If you call it after hitting an exception, it will automatically open an interactive debugging prompt at the point of the exception.

In [71]:
%debug

> <ipython-input-64-85edfa364281>(4)func1()
      2 
      3 def func1(a,b):
----> 4     return a/b
      5 
      6 def func2(x):

ipdb> print(a)
1
ipdb> print(b)
0
ipdb> quit


The interactive debugger allows much more than this, though–we can even step up and down through the stack and explore the values of variables there:

In [72]:
%debug

> <ipython-input-64-85edfa364281>(4)func1()
      2 
      3 def func1(a,b):
----> 4     return a/b
      5 
      6 def func2(x):

ipdb> up
> <ipython-input-64-85edfa364281>(9)func2()
      5 
      6 def func2(x):
      7     a = x
      8     b = x - 1
----> 9     return func1(a, b)

ipdb> print(x)
1
ipdb> down
> <ipython-input-64-85edfa364281>(4)func1()
      2 
      3 def func1(a,b):
----> 4     return a/b
      5 
      6 def func2(x):

ipdb> quit


This allows you to quickly find out not only what caused the error, but what function calls led up to the error.

If you'd like the debugger to launch automatically whenever an exception is raised, you can use the %pdb magic function to turn on this automatic behavior:

In [73]:
%xmode Plain

%pdb on
func2(1)


Exception reporting mode: Plain
Automatic pdb calling has been turned ON


ZeroDivisionError: division by zero

> <ipython-input-64-85edfa364281>(4)func1()
      2 
      3 def func1(a,b):
----> 4     return a/b
      5 
      6 def func2(x):

ipdb> print(b)
0
ipdb> quit


Finally, if you have a script that you'd like to run from the beginning in interactive mode, you can run it with the command **%run -d**, and use the next command to step through the lines of code interactively.

In [74]:
%run -d myscript.py

Breakpoint 1 at /home/nbuser/library/myscript.py:1
NOTE: Enter 'c' at the ipdb>  prompt to continue execution.
> /home/nbuser/library/myscript.py(1)<module>()
1---> 1 def square(x):
      2     """square a number"""
      3     return x ** 2
      4 
      5 for N in range(1, 4):

ipdb> c
1 squared is 1
2 squared is 4
3 squared is 9


There are many more available commands for interactive debugging than we've listed here; the following table contains a description of some of the more common and useful ones:

Command	Description

list	Show the current location in the file

h(elp)	Show a list of commands, or find help on a specific command

q(uit)	Quit the debugger and the program

c(ontinue)	Quit the debugger, continue in the program

n(ext)	Go to the next step of the program

<enter>	Repeat the previous command
    
p(rint)	Print variables

s(tep)	Step into a subroutine

r(eturn)	Return out of a subroutine

## Profiling and Timing Code

Donald Knuth famously quipped, "We should forget about small efficiencies, say about 97% of the time: premature optimization is the root of all evil."

But once you have your code working, it can be useful to dig into its efficiency a bit. 

IPython provides access to a wide array of functionality for this kind of timing and profiling of code. Here we'll discuss the following IPython magic commands:

**%time:** Time the execution of a single statement

**%timeit:** Time repeated execution of a single statement for more accuracy

**%prun:** Run code with the profiler

**%lprun:** Run code with the line-by-line profiler

**%memit:** Measure the memory use of a single statement

**%mprun:** Run code with the line-by-line memory profiler

**%timeit** line-magic and **%%timeit** cell-magic can be used to time the repeated execution of snippets of code.

Note that because this operation is so fast, %timeit automatically does a large number of repetitions. For slower commands, %timeit will automatically adjust and perform fewer repetitions:

In [77]:
%timeit sum(range(100))

1.31 µs ± 103 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


In [78]:
%%timeit
total = 0
for i in range(1000):
    for j in range(1000):
        total += i * (-1) ** j

379 ms ± 43.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


Sometimes repeating an operation is not the best option. For example, if we have a list that we'd like to sort, we might be misled by a repeated operation. Sorting a pre-sorted list is much faster than sorting an unsorted list, so the repetition will skew the result:

In [79]:
import random
L = [random.random() for i in range(100000)]
%timeit L.sort()

2.27 ms ± 122 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


For this, the **%time** magic function may be a better choice. It also is a good choice for longer-running commands, when short, system-related delays are unlikely to affect the result

In [80]:
L = [random.random() for i in range(100000)]
print("sorting an unsorted list:")
%time L.sort()

sorting an unsorted list:
CPU times: user 43.9 ms, sys: 0 ns, total: 43.9 ms
Wall time: 43.4 ms


In [82]:
print("sorting an already sorted list:")
%time L.sort()

sorting an already sorted list:
CPU times: user 3.27 ms, sys: 0 ns, total: 3.27 ms
Wall time: 3.36 ms


A program is made of many single statements, and sometimes timing these statements in context is more important than timing them on their own. Python contains a built-in code profiler and IPython offers a much more convenient way to use this profiler, in the form of the magic function **%prun**.

In [3]:
def sum_of_lists(N):
    total = 0
    for i in range(5):
        L = [j ^ (j >> i) for j in range(N)]
        total += sum(L)
    return total

In [84]:
%prun sum_of_lists(1000000)

The result is a table that indicates, in order of total time on each function call, where the execution is spending the most time. In this case, the bulk of execution time is in the list comprehension inside sum_of_lists. From here, we could start thinking about what changes we might make to improve the performance in the algorithm.

The function-by-function profiling of **%prun** is useful, but sometimes it's more convenient to have a line-by-line profile report. This is not built into Python or IPython, but there is a **line_profiler** package available for installation that can do this. 

Start by using Python's packaging tool, pip, to install the line_profiler package:

In [86]:
%load_ext line_profiler

In [87]:
%lprun -f sum_of_lists sum_of_lists(5000)

The information at the top gives us the key to reading the results: the time is reported in microseconds and we can see where the program is spending the most time.

Another aspect of profiling is the amount of memory an operation uses. This can be evaluated with another IPython extension, the **memory_profiler**.

In [1]:
%load_ext memory_profiler

The memory profiler extension contains two useful magic functions: the **%memit** magic (which offers a memory-measuring equivalent of %timeit) and the **%mprun** function (which offers a memory-measuring equivalent of %lprun).

In [4]:
%memit sum_of_lists(1000000)

peak memory: 125.32 MiB, increment: 78.09 MiB


We see that this function uses about 125 MB of memory.

For a line-by-line description of memory use, we can use the **%mprun** magic. Unfortunately, this magic works only for functions defined in separate modules rather than the notebook itself, so we'll start by using the **%%file** magic to create a simple module called mprun_demo.py, which contains our sum_of_lists function, with one addition that will make our memory profiling results more clear:

In [90]:
%%file mprun_demo.py
def sum_of_lists(N):
    total = 0
    for i in range(5):
        L = [j ^ (j >> i) for j in range(N)]
        total += sum(L)
        del L # remove reference to L
    return total

Writing mprun_demo.py


In [5]:
from mprun_demo import sum_of_lists
%mprun -f sum_of_lists sum_of_lists(1000)

Here the Increment column tells us how much each line affects the total memory budget: observe that when we create and delete the list L, we are adding about 25 MB of memory usage. This is on top of the background memory usage from the Python interpreter itself.